### PR TITLE
[MI-2314][server+webapp]: Made changes in plugin API for returning filter values and integrated UI subscripton filter flow with API.

### PR DESCRIPTION
**Pull request by avas27JTG**
- Made changes in plugin API for returning filter values based on different events for Repos
- Integrated UI subscription filter flow with API
-  Added conditional filters for Repos
- Added Area Path for Boards

### CODE REVIEW

**Review by raghavaggarwal2308 — APPROVED** [review]

File: webapp/src/containers/modals/SubscribeModal/filters/repos.tsx [hunk 1]
```diff
@@ -10,58 +10,65 @@ import usePluginApi from 'hooks/usePluginApi';
...
+        pluginConstants.pluginApiServiceConfigs.getSubscriptionFilters.apiServiceName,
+        getSubscriptionFiltersRequest as APIRequestPayload,
     );
-    const repositoryBranchesData = repositoryBranchesDataFromApi as ReposSubscriptionTargetBranchFilterResponse[] || [];
+    const filtersData = data as GetSubscriptionFiltersResponse || [];
 
-    const getTargetBranchOptions = () => ([{...filterLabelValuePairAll}, ...formLabelValuePairs('name', 'name', repositoryBranchesData)]);
+    const getRepositoryOptions = () => isSuccess && ([{...filterLabelValuePairAll}, ...formLabelValuePairs('displayValue', 'value', filtersData[subscriptionFiltersNameForRepos.repository], ['[Any]'])]);
+    const getTargetBranchOptions = () => isSuccess && ([{...filterLabelValuePairAll}, ...formLabelValuePairs('displayValue', 'value', filtersData[subscriptionFiltersNameForRepos.branch], ['[Any]'])]);
+    const getPullrequestCreatedByOptions = () => isSuccess && ([{...filterLabelValuePairAll}, ...formLabelValuePairs('displayValue', 'value', filtersData[subscriptionFiltersNameForRepos.pullrequestCreatedBy], ['[Any]'])]);
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/filters/repos.tsx:70]:
```suggestion
    const getPullRequestCreatedByOptions = () => isSuccess && ([{...filterLabelValuePairAll}, ...formLabelValuePairs('displayValue', 'value', filtersData[subscriptionFiltersNameForRepos.pullrequestCreatedBy], ['[Any]'])]);
```

File: webapp/src/containers/modals/SubscribeModal/filters/repos.tsx [hunk 2]
```diff
@@ -10,58 +10,65 @@ import usePluginApi from 'hooks/usePluginApi';
...
+    const getPullrequestReviewersContainsOptions = () => isSuccess && ([{...filterLabelValuePairAll}, ...formLabelValuePairs('displayValue', 'value', filtersData[subscriptionFiltersNameForRepos.pullrequestReviewersContains], ['[Any]'])]);
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/filters/repos.tsx:71]:
```suggestion
    const getPullRequestReviewersContainsOptions = () => isSuccess && ([{...filterLabelValuePairAll}, ...formLabelValuePairs('displayValue', 'value', filtersData[subscriptionFiltersNameForRepos.pullrequestReviewersContains], ['[Any]'])]);
```

File: webapp/src/utils/index.ts [hunk 3]
```diff
@@ -80,10 +80,14 @@ export const formLabelValuePair = (labelKey: string, valueKey: string, data: Rec
     return labelValuePair;
 };
 
-export const formLabelValuePairs = (labelKey: string, valueKey: string, data: Record<string, string>[]) => {
-    const labelValuePairs: LabelValuePair[] = [];
+export const formLabelValuePairs = (labelKey: string, valueKey: string, data: Record<string, string>[], excludeKeys: string[] = []) => {
+    let labelValuePairs: LabelValuePair[] = [];
     data.forEach((item) => labelValuePairs.push(formLabelValuePair(labelKey, valueKey, item)));
 
+    if (excludeKeys.length > 0) {
```
Comment (ayusht2810) [inline webapp/src/utils/index.ts:87]:
```suggestion
    if (excludeKeys.length) {
```

**Review by ayusht2810 — APPROVED** [review]

File: webapp/src/containers/modals/SubscribeModal/index.tsx [hunk 4]
```diff
@@ -21,16 +23,15 @@ import {getSubscribeModalState} from 'selectors';
...
 
-import './styles.scss';
-import {boardEventTypeOptions, repoEventTypeOptions} from 'pluginConstants/form';
-
-import {filterLabelValuePairAll} from 'pluginConstants/common';
-
 import ReposFilter from './filters/repos';
 
+import './styles.scss';
+import BoardsFilter from './filters/boards';
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/index.tsx:29]:
```suggestion
import BoardsFilter from './filters/boards';
import ReposFilter from './filters/repos';

import './styles.scss';
```

File: webapp/src/containers/modals/SubscribeModal/index.tsx [hunk 5]
```diff
@@ -239,6 +240,46 @@ const SubscribeModal = () => {
...
+            mergeResultName: name === filterLabelValuePairAll.value ? '' : name,
+        });
+
+    const handleSetPullRequestNotificationTypeFilter = (newValue: string, name?: string) =>
+        setSpecificFieldValue({
+            ...formFields,
+            notificationType: newValue,
+            notificationTypeName: name === filterLabelValuePairAll.value ? '' : name,
+        });
+    const handleSetAreaPathFilter = (newValue: string) =>
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/index.tsx:277]:
```suggestion
        });
        
    const handleSetAreaPathFilter = (newValue: string) =>
```

File: webapp/src/containers/modals/SubscribeModal/index.tsx [hunk 6]
```diff
@@ -275,15 +316,41 @@ const SubscribeModal = () => {
                                     ))
                                 }
                                 {
-                                    formFields.serviceType === pluginConstants.common.repos && (
+                                    formFields.serviceType === pluginConstants.common.boards && formFields.eventType && Object.keys(eventTypeBoards).includes(formFields.eventType) && (
+                                        <>
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/index.tsx:320]:
```suggestion
```

Comment (ayusht2810) [reply webapp/src/containers/modals/SubscribeModal/index.tsx:320]: i don't think we need this

File: webapp/src/containers/modals/SubscribeModal/index.tsx [hunk 7]
```diff
@@ -275,15 +316,41 @@ const SubscribeModal = () => {
...
+                                            <BoardsFilter
+                                                organization={organization as string}
+                                                projectId={projectID as string}
+                                                eventType={formFields.eventType || ''}
+                                                selectedAreaPath={formFields.areaPath || filterLabelValuePairAll.value}
+                                                handleSelectAreaPath={handleSetAreaPathFilter}
+                                                setIsFiltersError={setIsFiltersError}
+                                            />
+                                        </>
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/index.tsx:329]:
```suggestion
```

File: webapp/src/containers/modals/SubscribeModal/index.tsx [hunk 8]
```diff
@@ -275,15 +316,41 @@ const SubscribeModal = () => {
...
+                                    )
+                                }
+                                {
+                                    formFields.serviceType === pluginConstants.common.repos && formFields.eventType && Object.keys(eventTypeRepos).includes(formFields.eventType) && (
                                         <>
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/index.tsx:334]: same

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/client.go [hunk 9]
```diff
@@ -154,12 +154,43 @@ func (c *client) CreateSubscription(body *serializers.CreateSubscriptionRequestP
...
+				Branch:                       body.TargetBranch,
+				MergeResult:                  body.MergeResult,
+				PullRequestCreatedBy:         body.PullRequestCreatedBy,
+				PullRequestReviewersContains: body.PullRequestReviewersContains,
+			}
+		case constants.SubscriptionEventPullRequestUpdated:
+			payload.PublisherInputs = serializers.PublisherInputsRepos{
+				ProjectID:                    project.ProjectID,
+				Repository:                   body.Repository,
+				Branch:                       body.TargetBranch,
```
Comment (manojmalik20) [inline server/plugin/client.go:181]: You should add the common fields above the switch case, and then inside the switch , you can add fields according to the conditions.

File: server/serializers/subscriptions.go [hunk 10]
```diff
@@ -20,12 +20,18 @@ type PublisherInputsGeneric struct {
...
 
 type PublisherInputsRepos struct {
-	ProjectID  string `json:"projectId"`
-	Repository string `json:"repository"`
-	Branch     string `json:"branch"`
+	ProjectID                    string `json:"projectId"`
+	Repository                   string `json:"repository,omitempty"`
+	Branch                       string `json:"branch,omitempty"`
+	PullRequestCreatedBy         string `json:"pullrequestCreatedBy,omitempty"`
+	PullRequestReviewersContains string `json:"pullrequestReviewersContains,omitempty"`
```
Comment (manojmalik20) [inline server/serializers/subscriptions.go:31]:
```suggestion
	PullRequestCreatedBy         string `json:"pullRequestCreatedBy,omitempty"`
	PullRequestReviewersContains string `json:"pullRequestReviewersContains,omitempty"`
```

File: server/serializers/subscriptions.go [hunk 11]
```diff
@@ -51,21 +57,33 @@ type SubscriptionList struct {
...
+	Organization                     string `json:"organization"`
+	Project                          string `json:"project"`
+	EventType                        string `json:"eventType"`
+	ServiceType                      string `json:"serviceType"`
+	ChannelID                        string `json:"channelID"`
+	Repository                       string `json:"repository"`
+	RepositoryName                   string `json:"repositoryName"`
+	TargetBranch                     string `json:"targetBranch"`
+	PullRequestCreatedBy             string `json:"pullrequestCreatedBy"`
+	PullRequestReviewersContains     string `json:"pullrequestReviewersContains"`
```
Comment (manojmalik20) [inline server/serializers/subscriptions.go:69]:
```suggestion
	PullRequestCreatedBy             string `json:"pullRequestCreatedBy"`
	PullRequestReviewersContains     string `json:"pullRequestReviewersContains"`
```

File: server/serializers/subscriptions.go [hunk 12]
```diff
@@ -102,20 +120,31 @@ type CreateSubscriptionBodyPayload struct {
...
+	ChannelID                        string `json:"channelID"`
+	ChannelName                      string `json:"channelName"`
+	ChannelType                      string `json:"channelType"`
+	SubscriptionID                   string `json:"subscriptionID"`
+	CreatedBy                        string `json:"createdBy"`
+	TargetBranch                     string `json:"targetBranch"`
+	Repository                       string `json:"repository"`
+	RepositoryName                   string `json:"repositoryName"`
+	PullRequestCreatedBy             string `json:"pullrequestCreatedBy"`
+	PullRequestReviewersContains     string `json:"pullrequestReviewersContains"`
```
Comment (manojmalik20) [inline server/serializers/subscriptions.go:138]: Same.

File: server/serializers/subscriptions.go [hunk 13]
```diff
@@ -178,14 +207,20 @@ type Reviewer struct {
...
+	Organization                 string `json:"organization"`
+	Project                      string `json:"project"`
+	EventType                    string `json:"eventType"`
+	ServiceType                  string `json:"serviceType"`
+	ChannelID                    string `json:"channelID"`
+	MMUserID                     string `json:"mmUserID"`
+	TargetBranch                 string `json:"targetBranch"`
+	Repository                   string `json:"repository"`
+	PullRequestCreatedBy         string `json:"pullrequestCreatedBy"`
+	PullRequestReviewersContains string `json:"pullrequestReviewersContains"`
```
Comment (manojmalik20) [inline server/serializers/subscriptions.go:219]: Same.

File: webapp/src/components/card/subscription/index.tsx [hunk 14]
```diff
@@ -76,10 +76,30 @@ const SubscriptionCard = ({handleDeleteSubscrption, subscriptionDetails: {channe
...
                             <div className='card-chip-wrapper'>
                                 {
-                                    repositoryName && <Chip text={`Repository is ${repositoryName}`}/>
+
+                                    // Remove the extra character "/" from start and end of the area path string returned by the API
+                                    areaPath && <Chip text={`Area path - ${areaPath.substring(1, areaPath.length - 1)}`}/>
+                                }
+                                {
+                                    repositoryName && <Chip text={`Repository is: ${repositoryName}`}/>
+                                }
```
Comment (manojmalik20) [inline webapp/src/components/card/subscription/index.tsx:85]: I think writing these in one line would be better in terms of readability and will also reduce the LOC.

File: webapp/src/containers/modals/SubscribeModal/filters/boards.tsx [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    setIsFiltersError,
+}: BoardsFilterProps) => {
+    const {subscriptionFiltersNameForBoards, subscriptionFiltersForBoards} = pluginConstants.form;
+
+    const {
+        getApiState,
+        makeApiRequestWithCompletionStatus,
+    } = usePluginApi();
+
+    const getSubscriptionFiltersRequest = useMemo<GetSubscriptionFiltersRequest>(() => ({
```
Comment (manojmalik20) [inline webapp/src/containers/modals/SubscribeModal/filters/boards.tsx:36]: I don't thinkg using `useMemo` here gives us an advantage. What do you think?

File: webapp/src/containers/modals/SubscribeModal/filters/boards.tsx [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+    useEffect(() => {
+        if (isError && !isSuccess) {
+            setIsFiltersError(true);
+        } else {
+            setIsFiltersError(false);
+        }
+    }, [isLoading, isError, isSuccess]);
+
+    const getAreapathOptions = useCallback(() => (isSuccess ? ([{...filterLabelValuePairAll}, ...formLabelValuePairs('displayValue', 'value', filtersData[subscriptionFiltersNameForBoards.areaPath], ['[Any]'])]) : [pluginConstants.common.filterLabelValuePairAll]), [filtersData]);
```
Comment (manojmalik20) [inline webapp/src/containers/modals/SubscribeModal/filters/boards.tsx:66]:
```suggestion
    const getAreaPathOptions = useCallback(() => (isSuccess ? ([{...filterLabelValuePairAll}, ...formLabelValuePairs('displayValue', 'value', filtersData[subscriptionFiltersNameForBoards.areaPath], ['[Any]'])]) : [pluginConstants.common.filterLabelValuePairAll]), [filtersData]);
```

File: webapp/src/pluginConstants/form.ts [hunk 17]
```diff
@@ -263,3 +318,76 @@ export const subscriptionFilterEventTypeReposOptions = () => {
...
+
+// Repos subscription filters
+export const mergeResultOptons: LabelValuePair[] = [
+    {
+        value: 'Succeeded',
+        label: 'Merge successful',
+    },
+    {
+        value: 'Unsuccessful',
+        label: 'Merge Unsucessful - Reason: Any',
```
Comment (manojmalik20) [inline webapp/src/pluginConstants/form.ts:330]:
```suggestion
        label: 'Merge Unsuccessful - Reason: Any',
```
Fix this everywhere.

File: webapp/src/containers/modals/SubscribeModal/filters/repos.tsx [hunk 18]
```diff
@@ -10,58 +10,89 @@ import usePluginApi from 'hooks/usePluginApi';
...
+    selectedPullRequestCreatedBy: string
+    handleSelectPullRequestCreatedBy: (value: string, name?: string) => void
+    selectedPullRequestReviewersContains: string
+    handlePullRequestReviewersContains: (value: string, name?: string) => void
+    selectedPushedBy: string
+    handleSelectPushedBy: (value: string, name?: string) => void
+    selectedMergeResult: string
+    handleSelectMergeResult: (value: string, name?: string) => void
+    selectedNotificationType: string
+    handleSelectNotificationType: (value: string, name?: string) => void
```
Comment (manojmalik20) [inline webapp/src/containers/modals/SubscribeModal/filters/repos.tsx:29]: You can optimize this by passing two props only - 
Something like this - 
```
selectedFilter : {filterName: string, filterValue: string},
handleSelectFilter: (filterName: string, value: string, name?; string) => void
```

File: server/serializers/subscriptions.go (hunk 10 above)
Comment (avas27JTG) [reply server/serializers/subscriptions.go:31]: these cannot be modified as these are the expected key names for AZD

File: server/serializers/subscriptions.go (hunk 11 above)
Comment (avas27JTG) [reply server/serializers/subscriptions.go:69]: these cannot be modified as these are the expected key names for AZD

File: server/serializers/subscriptions.go (hunk 12 above)
Comment (avas27JTG) [reply server/serializers/subscriptions.go:138]: these cannot be modified as these are the expected key names for AZD

File: server/serializers/subscriptions.go (hunk 13 above)
Comment (avas27JTG) [reply server/serializers/subscriptions.go:219]: these cannot be modified as these are the expected key names for AZD

File: webapp/src/containers/modals/SubscribeModal/filters/boards.tsx (hunk 15 above)
Comment (avas27JTG) [reply webapp/src/containers/modals/SubscribeModal/filters/boards.tsx:36]: `selectedAreaPath` is a prop here which when changed in the parent this entire component will be re-rendered but we don't need that so, I think we should keep it

File: webapp/src/containers/modals/SubscribeModal/filters/repos.tsx (hunk 18 above)
Comment (avas27JTG) [reply webapp/src/containers/modals/SubscribeModal/filters/repos.tsx:29]: yes we can do something similar but here we will need to handle a few more things like in `handleSelectFilter: (filterName: string, value: string, name?; string) => void` we will need to target `name` field to put its value into the needed field only.

Have added a TODO for it, we can refactor it later after this release

File: server/plugin/api.go [hunk 19]
```diff
@@ -387,7 +400,24 @@ func (p *Plugin) handleGetSubscriptions(w http.ResponseWriter, r *http.Request)
...
+				subscriptionByProject[i].AreaPath <
+				subscriptionByProject[j].ChannelName+
+					subscriptionByProject[j].EventType+
+					subscriptionByProject[j].TargetBranch+
+					subscriptionByProject[j].PullRequestCreatedByName+
+					subscriptionByProject[j].PullRequestReviewersContainsName+
+					subscriptionByProject[j].PushedByName+
+					subscriptionByProject[j].MergeResultName+
+					subscriptionByProject[j].NotificationTypeName+
+					subscriptionByProject[j].AreaPath
```
Comment (manojmalik20) [inline server/plugin/api.go:420]: Why are these lines have more indentation than the above lines?

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**